target/linux/generic/pending-5.4/781-net-add-ndo-for-setting-the-iflink-property.patch

```diff
@@ -0,0 +1,92 @@
+From 846d8a31f9cc9048f58358770d20d3153918c93d Mon Sep 17 00:00:00 2001
+From: =?UTF-8?q?Marek=20Beh=C3=BAn?= <marek.behun@nic.cz>
+Date: Tue, 5 May 2020 19:28:53 +0200
+Subject: [PATCH] net: add ndo for setting the iflink property
+MIME-Version: 1.0
+Content-Type: text/plain; charset=UTF-8
+Content-Transfer-Encoding: 8bit
+
+In DSA the iflink value is used to report to which CPU port a given
+switch port is connected to. Since we want to support multi-CPU DSA, we
+want the user to be able to change this value.
+
+Add ndo_set_iflink method into the ndo strucutre to be a pair to
+ndo_get_iflink. Also create dev_set_iflink and call this from the
+netlink code, so that userspace can change the iflink value.
```

Comment:
How?

Reply:
It needs a change in iproute2, which I sent along with these patches.

But I don't think it will be accepted to upstream this way. The problem is that you cannot specify upstream port this way, only CPU port. So if you have multiple switches connected in cascade, like on Turris MOX, this approach falls short.

I need to look into this work once more and try to get it upstreamed.

Reply:
Do you mean connecting the 4 port and 8 port Ethernet modules?

Reply:
No, on Turris MOX there is only one CPU port. But imagine a situation where you have CPU and 2 switches. Each switch has 4 user ports, the CPU is connected to the switch1 via two ports, and switch1 is connected to switch2 via two ports:
```
 +--------+   +-------------------+   +----------------+
 |   eth0-------p0   switch1   p7-------p0  switch2    |
 | cpu    |   |                   |   |                |
 |   eth1-------p1 p2 p3 p4 p5 p6-------p1 p2 p3 p4 p5 |
 +--------+   +----|--|--|--|-----+   +----|--|--|--|--+
                   |  |  |  |              |  |  |  |
```
Such a situation (where there may be multiple DSA ports connecting 2 switches) is currently not handled at all and even with these patches you can't select via which DSA port packets should egress from switch2. But I have never seen a board with such schematics, so this is currently unimportant, fortunately.

Reply:
I see. In OpenWrt, there is no such device AFAIK. The motivation here is to bring back some of the speed for mvebu and mt7621. Both of these use swconfig in 19.07 (except Turris Omnia since it's...special).

Reply:
About speed on mt7621 (mt7530). It's low at that moment. Only 500Mbite/s. Not sure those patches help much. I'm using them right now. Tested with iperf3 Router<->PC. Is it possible to get 1Gbite/s on DSA or I'm wrong somewhere? Maybe it's better to use the old driver on such weak devices?

Reply:
mt7621 has one CPU port in the switch. Some devices have two, but those are very rare.

Reply:
@Vladdrako
> iperf3 Router<->PC

Do not run iperf3 on your router, as it will starve your router's CPU, and flow offloading won't work that way.

@neheb
>  Some devices have two, but those are very rare.

MT7621's second GMAC can be used as the second CPU port, unless the RGMII2 pins are used for something else.
It's not _rare_ at all. It's just the previous swconfig driver does not use the second GMAC.

Reply:
That's still the case. I don't know of any mt7621 device that defines the second ethernet interface in DTS.

Reply:
@neheb
Take a look at the vendor driver's kernel-3.4.x.config
https://github.com/hanwckf/rt-n56u/tree/master/trunk/configs/boards
`CONFIG_RAETH_BOTH_GMAC` means both GMACs can be used.

And FYI, even if the current MT7530 DSA does not support multi CPU ports, you can "hard-wire" port 0 or port 4 to the second GMAC by modifying DTS.

Credit @vDorst

Reply:
@dengqf6 In client mode too. 500-550Mbit/s maximum. Or is there any other way to test the speed?

Reply:
> In client mode too

Whatever mode it is, you should not run it on your router.

You will need two **PC**s, one connected to WAN as server and the other to LAN as client. Enable hardware offloading on the router.